### DISCUSSION TITLE
make two related IDS unique

### DISCUSSION THREAD

**Discussion by fuedgabriel** [prisma/prisma#23270] [Q&A] [answered]
### Question

Would there be a solution to this problem that relates the two ids and makes them unique, but only if there are these two ids, if it is the same id and a different one, it is no longer unique. This via schem.prisma

I had the same problem, since the table is not unique. Code below with the problem.

```js
const watchedAlreadyExists = await prisma.watched.findUnique(
    {
        where: {
            userId: userId,
            episodeId: episodeId
        }
    }
);
```

```
The type '{ userId: string; episodeId: string; }' cannot be assigned to type 'WatchedWhereUniqueInput'.
   The type '{ userId: string; episodeId: string; }' cannot be assigned to type '{ id: string; } & { id?: string | undefined; AND?: WatchedWhereInput | WatchedWhereInput[] | undefined; OR?: WatchedWhereInput[] | undefined; ... 9 more ...; episode?: (Without<...> & EpisodesWhereInput) | ... 1 more ... | undefined; }'.
     Property 'id' is missing in type '{ userId: string; episodeId: string; }', but it is mandatory in the type '{ id: string; }'.ts(2322)
index.d.ts(6518, 5): The expected type comes from the 'where' property, which is declared here in the type '{ select?: WatchedSelect<DefaultArgs> | null | undefined; include?: WatchedInclude<DefaultArgs> | null | undefined; where: WatchedWhereUniqueInput; }
```

Solution:

`const watchedAlreadyExists = await prisma.watched.findFirst({
        where: {
            userId: userId,
            episodeId: episodeId
        }
    }
);`

The solution is to place a findFirst, but there will not be another element like this in the table. I think this solution is due to bad programming practices

![image](https://github.com/prisma/prisma/assets/25815971/a2616f7e-19f5-492a-afc5-572d0c664016)

![image](https://github.com/prisma/prisma/assets/25815971/e15327e4-1a47-4ae7-b0a3-e14f20ab4a93)

- OS: 
- Database:  v20.11.0
- yarn run v1.22.21
- 
  "devDependencies": {
    "@types/node": "^20.11.19",
    "prisma": "^5.9.1",
    "ts-node": "^10.9.2",
    "tsx": "^4.7.1",
    "typescript": "^5.3.3"
  },

**Comment by ludralph** [comment, marked as answer]
Hi @fuedgabriel 👋 

Thank you for raising this question.

You are encountering a type mismatch issue when trying to use `findUnique` with multiple fields (userId and episodeId) that are not uniquely identifying a record by themselves in the Prisma schema. The error message indicates that Prisma expects a `WatchedWhereUniqueInput` type which typically requires a single unique identifier, such as an `id`.

To address the issue of finding a record based on a combination of `userId` and `episodeId` that are not individually unique but together should identify a unique record, you correctly identified that using `findFirst` is a solution. This approach is valid when you want to find a record based on criteria that do not conform to the unique constraints expected by `findUnique`.
Here's the solution you mentioned
```js
const watchedAlreadyExists = await prisma.watched.findFirst({
    where: {
        userId: userId,
        episodeId: episodeId
    }
});
```

While you expressed concern that this might be due to bad programming practices, it's important to note that `findFirst` is a legitimate method provided by Prisma for cases where the criteria used for finding records do not meet the strict requirements of `findUnique`. 

`findFirst` will return the first record that matches the criteria specified in the `where` clause, which in practice can be used similarly to `findUnique` when you are confident that the combination of criteria (in this case, `userId` and `episodeId`) will only match a single record. For ensuring uniqueness on a combination of fields directly in your Prisma schema, you can use the [`@@unique` ](https://www.prisma.io/docs/orm/reference/prisma-schema-reference#unique-1) attribute on the model. This would enforce the uniqueness constraint at the database level. For example, here's how you could define such a constraint in your Prisma schema:

```prisma
@@unique([userId, episodeId], name: "unique_user_episode")
```

This would enforce that the combination of `userId` and `episodeId` is unique across all records in the table, allowing you to confidently use `findUnique` with this combination of fields.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by fuedgabriel** [reply]
Thank you very much!! great explanation, literally everything perfect!! @ludralph
